src/Compilers/Core/Portable/Collections/SmallDictionary.cs

```diff
@@ -29,11 +33,12 @@ namespace Microsoft.CodeAnalysis
     /// then regular Dictionary is a better choice.
     /// </summary>
     internal sealed class SmallDictionary<K, V> : IEnumerable<KeyValuePair<K, V>>
+        where K : notnull
     {
-        private AvlNode _root;
+        private AvlNode? _root;
         public readonly IEqualityComparer<K> Comparer;
 
-        public static readonly SmallDictionary<K, V> Empty = new SmallDictionary<K, V>(null);
+        public static readonly SmallDictionary<K, V> Empty = new SmallDictionary<K, V>(EqualityComparer<K>.Default);
```

Comment:
💭  It's customary for constructors to allow null comparers. Any reason not to leverage that here?

Reply:
This is also a case where NRT is pointing out that this behavior is wrong. Consider the following code

```
SmallDictionary<int, int>.Empty.Add(42, 13);
```

This will fail with a `NullReferenceException` as the call to `GetHashCode` assumes `Comparer` is a non-null value. Allowing `null` here would require re-engineering this type a bit. 

---
In reply to: [356224843](https://github.com/dotnet/roslyn/pull/40268#discussion_r356224843) [](ancestors = 356224843)

Reply:
We could argue that this behavior is "By Design". Essentially the exception here is useful in enforcing that `Empty` is never mutated. If that is the argument though then I'd like to do it in a much more declarative way. 

---
In reply to: [356281462](https://github.com/dotnet/roslyn/pull/40268#discussion_r356281462) [](ancestors = 356281462,356224843)

Reply:
I mean the constructor provides the default value if null was specified.

Reply:
That is reasonable ... but it would also break the behavior here. Either way I think this type needs to envolve in some way to make this implicit contract explicit

---
In reply to: [356297727](https://github.com/dotnet/roslyn/pull/40268#discussion_r356297727) [](ancestors = 356297727)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/40344](https://github.com/dotnet/roslyn/issues/40344) to track this pattern.

---
In reply to: [356315553](https://github.com/dotnet/roslyn/pull/40268#discussion_r356315553) [](ancestors = 356315553,356297727)